src/runtime.js

```diff
@@ -408,36 +411,156 @@ function registerServiceForDoc(ampdoc, name, opt_ctor, opt_factory) {
-  // TODO(dvoytenko): switch this method to accept the actual shadow document
-  // and attach shadow in runtime.
-  dev.fine(TAG, 'Attach shadow root:', shadowRoot, extensionIds);
+function prepareAndAttachShadowDoc(global, extensions, hostElement, doc, url) {
+  dev.fine(TAG, 'Attach shadow doc:', doc);
   const ampdocService = ampdocFor(global);
-  const ampdoc = ampdocService.getAmpDoc(shadowRoot);
 
+  hostElement.style.visibility = 'hidden';
+  const shadowRoot = hostElement.createShadowRoot();
   shadowRoot.AMP = {};
+  shadowRoot.AMP.url = url;
+
+  const ampdoc = ampdocService.getAmpDoc(shadowRoot);
+  dev.fine(TAG, 'Attach to shadow root:', shadowRoot, ampdoc);
 
   // Install runtime CSS.
   installStylesForShadowRoot(shadowRoot, cssText,
       /* opt_isRuntimeCss */ true, /* opt_ext */ 'amp-runtime');
 
-  // Install services.
+  // Instal doc services.
   installAmpdocServices(ampdoc);
+
+  // Install extensions.
+  const extensionIds = mergeShadowHead(global, extensions, shadowRoot, doc);
+
+  // Apply all doc extensions.
   installExtensionsInShadowDoc(extensions, ampdoc, extensionIds);
 
+  // Append body.
+  if (doc.body) {
+    const body = global.document.importNode(doc.body, true);
+    body.classList.add('amp-shadow');
+    body.style.position = 'relative';
+    shadowRoot.appendChild(body);
+  }
+
+  // TODO(dvoytenko): find a better and more stable way to make content visible.
+  // E.g. integrate with dynamic classes. In shadow case specifically, we have
+  // to wait for stubbing to complete, which may take awhile due to importNode.
+  global.setTimeout(function() {
+    hostElement.style.visibility = 'visible';
+  }, 50);
+
   dev.fine(TAG, 'Shadow root initialization is done:', shadowRoot, ampdoc);
   return shadowRoot.AMP;
 }
 
 
 /**
+ * Processes the contents of the shadow document's head.
+ * @param {!Window} global
+ * @param {!./service/extensions-impl.Extensions} extensions
+ * @param {!ShadowRoot} shadowRoot
+ * @param {!Document} doc
+ * @return {!Array<string>}
+ */
+function mergeShadowHead(global, extensions, shadowRoot, doc) {
+  const extensionIds = [];
+  if (doc.head) {
+    const parentLinks = {};
+    childElementsByTag(global.document.head, 'link').forEach(link => {
+      const href = link.getAttribute('href');
+      if (href) {
+        parentLinks[href] = true;
+      }
+    });
+
+    for (let n = doc.head.firstElementChild; n; n = n.nextElementSibling) {
+      const tagName = n.tagName;
+      const name = n.getAttribute('name');
+      const rel = n.getAttribute('rel');
+      if (n.tagName == 'TITLE') {
+        shadowRoot.AMP.title = n.textContent;
+        dev.fine(TAG, '- set title: ', shadowRoot.AMP.title);
+      } else if (tagName == 'META' && n.hasAttribute('charset')) {
+        // Ignore.
+      } else if (tagName == 'META' && name == 'viewport') {
+        // Ignore.
+      } else if (tagName == 'META') {
+        // TODO(dvoytenko): copy other meta tags.
```

Comment:
The shell does have the whole DOM so it can do it anyway. However, we use META for services and they must stay in the ampdoc's DOM. E.g. AMP Access accepts the server state as a META tag and when AMP Access is migrated to be ampdoc service it should be able to find this info here.